crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let copied = obs.clone();
+    Box::into_raw(Box::new(copied))
+}
+
+/// @ingroup QkSparseObservable
+/// Compare two observables for equality.
+///
+/// Note that this does not compare mathematical equality, but data equality. This means
+/// that two observables might represent the same observable but not compare as equal.
+///
+/// @param obs A pointer to one observable.
+/// @param other A pointer to another observable.
+///
+/// @return ``true`` if the observables are equal, ``false`` otherwise.
+///
+/// # Example
+///
+///     QkSparseObservable *observable = qk_obs_identity(100);
+///     QkSparseObservable *other = qk_obs_identity(100);
+///     bool are_equal = qk_obs_equal(observable, other);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` or ``other`` are not valid, non-null pointers to
+/// ``QkSparseObservable``\ s.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_equal(
+    obs: *const SparseObservable,
+    other: *const SparseObservable,
+) -> bool {
+    // SAFETY: Per documentation, the pointers are non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let other = unsafe { const_ptr_as_ref(other) };
+
+    obs.eq(other)
+}
+
+/// @ingroup QkSparseObservable
+/// Return a string representation of a ``SparseObservable``.
+///
+/// @param obs A pointer to the ``SparseObservable`` to get the string for.
+///
+/// @return A pointer to a nul-terminated char array of the string representation for ``obs``
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     char *string = qk_obs_str(obs);
+///     qk_free_obs_str(string);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+///
+/// The string must not be freed with the normal C free, you must use ``qk_free_obs_str`` to
+/// free the memory consumed by the String. Not calling ``qk_free_obs_str`` will lead to a
+/// memory leak.
+///
+/// Do not change the length of the string after it's returned, although values can be mutated
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_str(obs: *const SparseObservable) -> *mut c_char {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let string: String = format!("{:?}", obs);
+    CString::new(string).unwrap().into_raw()
+}
+
+/// @ingroup QkSparseObservable
+/// Free a string representation of a ``SparseObservable``.
+///
+///  @param str A pointer to the returned string representation from ``qk_obs_str``
+///
+/// # Safety
+///
+/// Behavior is undefined if ``str`` is not a pointer returned by ``qk_obs_str``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_free_obs_str(string: *mut c_char) {
```

Comment:
This arguably breaks the naming convention of `qk_<obj>_<fn>` - it should be probably be `qk_obs_free_str`?

Reply:
Since this frees any string, maybe we could just have this be `qk_str_free`?

Reply:
We can unify them to that - it's not dependent.